Feature/ci prevent new cppcheck issues

rmelotte opened the pull request:
With the huge amount of issues cppcheck reports already it will take a lot of time to solve them all, and it could be useful to still prevent new issues like we do for PEP8.

Since we've never been block by "less severe" cppcheck issues before, this means we will potentially have to tweak cppcheck from time to time for issues we think are not relevant, and PR might get block by it in the meantime.

rmelotte committed:
ci: add a file with existing cppcheck issues

Add a file with the current cppcheck issues.

It will later be used to check that no new issues are introduced.

It uses a custom cppcheck template:
'{file}: {severity}: {message} [{id}]{code}'

The template voluntarily excludes line and column numbers, to avoid
the need to update the list of known issue every time a line number
changes (adding a new include for example would change all the line
numbers that follow).

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
ci: add a script that catches new cppcheck issues

We cannot fix all cppcheck issues at once because there are too many.

In the meantime, what we can do is to avoid new issues to be
introduced.

Add a script that checks whether the current cppcheck_results.txt file
contains new issues or not.

It will first generate a list of known issue in the same format as the
known issues (so excluding line and column number among other things).

For each issues in cppcheck_results.txt, it will then check if it
exists in ci/cppcheck/cppcheck_existing_issues.txt.
If not, it will print an error messages and display the corresponding
cppcheck issue.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
.gitlab-ci.yml: add a job to check for new issues

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte commented:
~~We could also make the job "allowed to fail" for some time, until we are confident it only reports relevant issues.~~
As @arnout mentioned during the sync meeting, making the job "allowed to fail" will encourage people to **not** check its status, which means it won't be taken into account in the end.

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -0,0 +1,26 @@
+    err "run $rootdir/tools/docker/static-analysis/cppcheck.sh first"
+fi
+
+mapfile -t issues < <(grep -E ".+:[0-9]+:[0-9]+: .+: .+ \[.+\]$"  "$rootdir/cppcheck_results.txt")
```
TIL: the mapfile command. Interesting!

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -0,0 +1,26 @@
+
+
+if [ ! -f "$rootdir/cppcheck_results.txt" ] ; then
+    err "run $rootdir/tools/docker/static-analysis/cppcheck.sh first"
```
Missing exit after printing the error.

arnout commented on ci/cppcheck/cppcheck-new-issues.sh:
```diff
@@ -0,0 +1,26 @@
+mapfile -t issues < <(grep -E ".+:[0-9]+:[0-9]+: .+: .+ \[.+\]$"  "$rootdir/cppcheck_results.txt")
+
+status=0
+for i in "${issues[@]}"
```
(nitpick) As happy as I am with learning the mapfile command, here it really doesn't bring anything over the traditional (and POSIX)

```
grep -E ... | \
    while read -r i; do
```

arnout reviewed (approved):
I will approve this PR, but I'm afraid we're going to revert it pretty quickly...

The problem is that the issue matching happens based on the line numbers. For example, if we add an include to db.cpp, all the errors that it currently gives will shift by one line. So we need to add an 800-line diff to cppcheck_existing_issues.txt to handle that... The result will be that people blindly overwrite cppcheck_existing_issues.txt so you still don't see a new issue.

To work around it, we could use a different output format where the code is on the same line as the rest, and where we don't have line and column numbers.

For now though, I'd merge as is. We can still fix it if it turns out to be an actual problem.

rmelotte commented:
Indeed the line number is a pretty annoying issue, I'm glad you mentioned it.
In the end it was easy to change the format of the reported issues since we're doing this in a separate job anyway.
So I did exactly what you suggested: removed line and column numbers and added the code itself.
